### PR TITLE
feat: add button components

### PR DESCRIPTION
**Pull request by alexlafroscia**
Adds Tailwind components for our Fluid button styles. Check the Percy snapshots for an idea of what everything looks like implemented!

~~These new classes use the same naming convention as the old button classes. After thinking about it a bit, I couldn't really come up with a better system, as the existing on is already pretty good and works well for composability. This will also allow us to just remove the existing CSS definition for our buttons at the point that we upgrade `front-end` to use a version of Tailwind that includes this, and all of our current Tailwind buttons will "magically" begin using the Tailwind definition instead.~~

The Fluid button styles are meant to compose and build on one another.

You start by rendering a "basic" button like so

```html
<button class="fluid-button">
  A Basic Button
</button>
```

This should be the default choice based on size and style, and thus has no additional configuration.

You can set the "type" of Fluid button by applying one of our `type:`-prefixes classes. This syntax is borrowed from Tailwind itself, which uses this type of colon-delimited CSS class for variants within it's system. We can borrow this syntax for a less verbose means of setting the type.

```html
<button class="fluid-button type:primary">
  A Primary Button
</button>
```

This can be further extended to handle button sizes as well

```html
<button class="fluid-button type:primary size:large">
  A Large Primary Button
</button>
```

Rendering an icon-only button requires placing an SVG inside the button -- it is automatically sized correctly, no matter which size modifier is provided

```html
<button class="fluid-button size:lg" aria-label="Some Description">
  <svg .... />
</button>
```

Placing text and an icon is also supported -- you just need to wrap the text in a `<span>` ☹️ 

```html
<button class="fluid-button size:lg">
  <span>Some Description</span>
  <svg .... />
</button>
```

It is impossible to detect a bare text node in CSS, so even if text and an icon are included in the HTML, the `svg` still considers itself the only child element. Adding the `<span>` allows the `svg` to add the right amount of spacing between it and the text.

## Terminology

I had to define some vocabulary in here that I think we should more widely adopt to ensure everyone is on the same page

<details>
  <summary>Type</summary>
  <p>

The button style to use. Primary, destructive, etc.

The design system documentation refers to this as the "state", but this is confusing because button states normally refer to things like a button being disabled, active, focused, etc.

</p></details>

<details>
  <summary>State</summary>
  <p>

Corresponds to whether a button is focused, disabled, etc.

In addition to the pseudo-selectors for each of these states, a class is also generated for each for instances where, for whatever reason, the normal CSS pseudo-class cannot be used. For example, we may want to style an anchor (`<a>`) like a disabled button, but it has no `:disabled` CSS API.

</p></details>

## Implementation Details

The bit of code that generates the base styles for the buttons looks bit strange:

https://github.com/movableink/tailwind-config/compare/fluid-button?expand=1#diff-91d6b753b85474d2728966217ce6d7aaR22

This is making use of advanced CSS selectors in order to share those basic styles with _all_ Fluid buttons, without that CSS needing to be repeated in each of the definitions of the button types.

## To-Do

- [x] Button "type" components
- [x] Button "states"
- [x] Button sizes
- [x] Icon-only buttons
- [x] Icon-with-text buttons
- [x] Fix styling in FireFox (complex attribute selector was slightly "off" in FireFox; being more strict on the syntax made it work there 😌)

## Open Questions/Things to Resolve

<details>
  <summary>Resolved Questions</summary>
  <p>

***

- [x] The Sketch file that defines out button states does not really take border thickness into account. The buttons appear to be 28px high by default, but the height is calculated as

    > line height + (vertical padding * 2)

    This does not allow for the border at all, which is by default `1px`

    This is additionally problematic because the focus state uses a border of `2px`, meaning the button is taller when it is focused. 
- [x] The "outline" button type uses a transparent background, from what @keser and I talked about (not white). It is unclear how exactly the disabled state should operate, as normally the color appears lighter.

    Options:
    1. Use a transparent background color that's something like `rgba(255, 255, 255, 0.1)` to create a lightening of whatever the existing background color is behind the button
    2. Just do nothing
    3. Something else?
- [x] I think we need to look at the `:active` state and work on the way it interacts with the `:focus` state.

  By default, the `:active` and `:focus` states are probably both going to be present when an element is clicked on (which makes sense; it is both focused _and_ clicked at the same time). However, our `:active` styles are not an extension of our `:focus` styles. Writing the CSS to match the design, we have to manually override the `:focus` styles to set them back to the initial values.

  This is fine, implementation wise, but creates a bit of an awkward UX

  Kapture 2020-05-26 at 16 12 37

  The button is depressed-looking (with the `inset` shadow) when clicked, but upon release, enters _only_ the `:focus` state and gets the border. It just looks a bit weird, especially because the `:focus` border is thicker than the "normal" one.

  There is a similar issue between `:active` and `:hover` -- many `:active` items will _also_ be hovered, so we need to decide if the `:active` state should allow the `:hover` states to _also_ be present, or if they should explicitly be overridden.
- [x] Should we disallow text to break to multiple lines within a button?
- [x] Does the `type:foo` syntax feel weird for providing the "type" of button, when Tailwind uses this for variants? The types and sizes sort of are variants, and from that perspective I think I like it, but there are certainly alternatives (cc: @nicksteffens @spencer516 @cwervo)
  1. Use the `btn` class and attach modifiers to it (`btn-sm`, `btn-primary`)
      * I am concerned that `btn` may be used elsewhere in our app already (or by third-party dependencies 🤢)
      * This is the naming that GitHub uses in their design system
  2. Use the existing [`fluid-button` class naming](https://fluid.movableink-dmz.com/docs/components/fluid-button)
      * This is great when you want to reach for just at type (as in, `fluid-button--primary` is shorter than `fluid-button type:primary`
      * This gets really repetitive when you have a size as well (`fluid-button--primary fluid-button--small`)
      * The naming convention does not imply what the default button style is (you _always_ must explicitly choose a "type")
- [x] Are the buttons the right height? The rendered sizes don't _quite_ match what appears in Sketch (likely due to Sketch not really handling borders the same way that the web does)
    * Compact -- `22px`
    * Small -- `26px`
    * Base -- `30px`
    * Large -- `34px`
    * Extra Large -- `46px`
- [x] How should we name the size classes? With a short-hand (like `sm`) or writing it all out (like `small`)
    * A shorthand is less descriptive
    * Something like `extra-large` is _pretty_ long compared to `xl`

***

</p></details>

### CODE REVIEW

File: plugins/components/buttons.js [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+        boxShadow: 'none',
+      }),
+
+      ...focused({
+        outline: 'none',
+      }),
+    },
+
+    /** === Basic Button === **/
+    [`.fluid-button:not(*[class*=${e('type:')}])`]: {
```
Comment (alexlafroscia) [inline plugins/components/buttons.js:83]: This thing is pretty gross, but exists for a good reason!

Originally I had implemented `.fluid-button` with the base styles for all buttons _plus_ the styles for the "basic button". However, that left a number of locations where button types has to add additional CSS to themselves just to override the "basic" styles that they never cared about in the first place.

This selector says, essentially

> Match any `.fluid-button` that does not have a `type:` class on it

Which allows us to separate out the "shared" and "basic" styles again, and avoid any of the other "type" styles needing to override the styles from the "basic" button.

Comment (spencer516) [reply plugins/components/buttons.js:83]: 🤯 — is it worth capturing this explanation as a comment in the code itself?

Comment (cwervo) [reply plugins/components/buttons.js:83]: +1 to adding this as a comment in the CSS itself, as this is pretty gnarly to figure out on your own (which I just spent a few minutes doing prior to see this GH comment lol)

**Comment by cwervo** [comment]
> Does the type:foo syntax feel weird for providing the "type" of button, when Tailwind uses this for variants? The types and sizes sort of are variants, and from that perspective I think I like it, but there are certainly alternatives
- re: option _i._ `btn-*` isn't horrible but probably clashes with something, somewhere, & also just doesn't skim in CSS as well as `type:`
- I'm against option _ii._ because of the last two bullets you listed:
  > - This gets really repetitive when you have a size as well (fluid-button--primary fluid-button--small)
  >  - The naming convention does not imply what the default button style is (you always must explicitly choose a "type")

I think `type:*` works super well!

**Comment by alexlafroscia** [comment]
@cwervo thanks for the feedback on the modifier syntax!

I posted it to Twitter as well and got similar feedback on the `type:xxx` syntax not being too weird or anything

https://twitter.com/alexlafroscia/status/1265715752444796928?s=20

File: plugins/components/buttons.js (hunk 1 above)
Comment (alexlafroscia) [reply plugins/components/buttons.js:103]: Good call. Done!

**Comment by alexlafroscia** [comment]
This should be ready for a real review -- still looking for input on whether we like `sm/lg` etc or should write out `small/large` etc. in full for the size settings.

**Comment by alexlafroscia** [comment]
> When we update the version of tailwind into our fluid repo, are the buttons going to immediately replace the existing fluid-button styles?

Depends on what you mean by “immediately” 😅 I am thinking it’s going to go something like this:

1. Upgrade the Tailwind package to include the button styles
2. Take a pass at the Rails app to remove any use of the BEM-style buttons and replace with this API instead
    * This happens first because removing the BEM-style buttons from Fluid will make them inaccessible to Rails; it needs to migrate off of that API first
3. Land a big ol’ PR to the `front-end` that swaps out it’s BEM button usage for our new button API

All of this I intend to start today (as soon as this PR lands) and work on it Tuesday/Wednesday.

Note that the original way this was implemented — continuing to use the BEM naming style — _would_ have allowed us to just drop the BEM implementation, install this one, and call it a day. But since we changed the API we’ll need a manual migration in a very similar way to how we migrate from the in-house utilities in favor of the Tailwind ones.

**Comment by alexlafroscia** [comment]
Totally agree @nicksteffens -- @keser and I also talked about it and he mentioned the same reasoning for using those abbreviations. We're already using them elsewhere!

**Pull request merged**